derp/derp_server.go

```diff
@@ -889,6 +981,96 @@ func (c *sclient) sendPacket(srcKey key.Public, contents []byte) (err error) {
 	return err
 }
 
+// AddPacketForwarder registers fwd as a packet forwarder for dst.
+// fwd must be comparable.
+func (s *Server) AddPacketForwarder(dst key.Public, fwd PacketForwarder) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.clientsElsewhere == nil {
+		s.clientsElsewhere = map[key.Public]PacketForwarder{}
+	}
+	if prev, ok := s.clientsElsewhere[dst]; ok {
+		if prev == fwd {
+			// Duplicate registration of same forwarder. Ignore.
+			return
+		}
+		if m, ok := prev.(multiForwarder); ok {
+			if _, ok := m[fwd]; !ok {
+				// Duplicate registration of same forwarder in set; ignore.
+				return
+			}
+			m[fwd] = m.maxVal() + 1
+			return
+		}
+		// Otherwise, the existing value is not a set and not a dup, so make it a set.
+		fwd = multiForwarder{
+			prev: 1,
+			fwd:  2,
+		}
+	}
+	s.clientsElsewhere[dst] = fwd
+}
+
+// RemovePacketForwarder removes fwd as a packet forwarder for dst.
+// fwd must be comparable.
+func (s *Server) RemovePacketForwarder(dst key.Public, fwd PacketForwarder) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	v, ok := s.clientsElsewhere[dst]
+	if !ok {
+		return
+	}
+	if m, ok := v.(multiForwarder); ok {
+		if len(m) < 2 {
+			panic("unexpected")
+		}
+		delete(m, fwd)
+		// If fwd was in m and we no longer need to be a
+		// multiForwarder, replace the entry with the
+		// remaining PacketForwarder.
+		if len(m) == 1 {
+			var remain PacketForwarder
+			for k := range m {
+				remain = k
+			}
+			s.clientsElsewhere[dst] = remain
+		}
+		return
+	}
+	if v != fwd {
+		// Delete of an entry that wasn't in the map.
+		return
+	}
+	delete(s.clientsElsewhere, dst)
+}
+
+// multiForwarder is a set of forwarders. It's used in the rare cases that a client
+// is connected to multiple DERP nodes in a region. The value is unique connection
+// number; the lowest one has been seen the longest. It's used to make sure we
+// forward packets consistently to the same node and don't pick randomly.
+type multiForwarder map[PacketForwarder]uint8
+
+func (m multiForwarder) maxVal() (max uint8) {
+	for _, v := range m {
+		if v > max {
+			max = v
+		}
+	}
+	return
+}
+
+func (m multiForwarder) ForwardPacket(src, dst key.Public, payload []byte) error {
```

Comment:
If we expect the number of forwarders to be small, maybe make this a struct with an array? That'd be faster to scan linearly, and also saves having to find the latest connection each time - it's always at the back of the array.

Reply:
Yeah, suppose I could. But OTOH I'm not really worried about the performance of looking up in them because I expect these maps to not last very long, just while clients are moving around.

Reply:
.... if ever. It's possible this is never used. I should add a varz!